pramen/core/src/main/scala/za/co/absa/pramen/core/runner/splitter/ScheduleStrategyIncremental.scala

```diff
@@ -0,0 +1,100 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package za.co.absa.pramen.core.runner.splitter
+
+import za.co.absa.pramen.api.status.{MetastoreDependency, TaskRunReason}
+import za.co.absa.pramen.core.bookkeeper.Bookkeeper
+import za.co.absa.pramen.core.bookkeeper.model.DataOffsetAggregated
+import za.co.absa.pramen.core.pipeline
+import za.co.absa.pramen.core.pipeline.TaskPreDef
+import za.co.absa.pramen.core.runner.splitter.ScheduleStrategyUtils.{log, _}
+import za.co.absa.pramen.core.schedule.Schedule
+
+import java.time.LocalDate
+
+class ScheduleStrategyIncremental(lastOffsets: Option[DataOffsetAggregated], hasInfoDateColumn: Boolean) extends ScheduleStrategy {
+  private val log = org.slf4j.LoggerFactory.getLogger(this.getClass)
+
+  override def getDaysToRun(
+                             outputTable: String,
+                             dependencies: Seq[MetastoreDependency],
+                             bookkeeper: Bookkeeper,
+                             infoDateExpression: String,
+                             schedule: Schedule,
+                             params: ScheduleParams,
+                             initialSourcingDateExpr: String,
+                             minimumDate: LocalDate
+                           ): Seq[TaskPreDef] = {
+    val dates = params match {
+      case ScheduleParams.Normal(runDate, trackDays, _, _, _) =>
+        val infoDate = evaluateRunDate(runDate, infoDateExpression)
+        log.info(s"Normal run strategy: runDate=$runDate, infoDate=$infoDate")
+
+        val runInfoDays = if (hasInfoDateColumn) {
+          lastOffsets match {
+            case Some(lastOffset) =>
+              if (lastOffset.maximumInfoDate.isBefore(infoDate)) {
+                val startDate = if (trackDays > 1) {
+                  val trackDate = infoDate.minusDays(trackDays - 1)
+                  val date = if (trackDate.isAfter(lastOffset.maximumInfoDate))
+                    trackDate
+                  else
+                    lastOffset.maximumInfoDate
+                  log.warn(s"Last ran day: ${lastOffset.maximumInfoDate}. Tracking days = '$trackDate'. Catching up from '$date' until '$infoDate'.")
+                  date
+                } else {
+                  log.warn(s"Last ran day: ${lastOffset.maximumInfoDate}. Catching up data until '$infoDate'.")
+                  lastOffset.maximumInfoDate
+                }
+
+                val potentialDates = getInfoDateRange(startDate, infoDate, "@runDate", schedule)
+                potentialDates.map(date => {
+                  TaskPreDef(date, TaskRunReason.New)
+                })
+              } else {
+                Seq(TaskPreDef(infoDate, TaskRunReason.New))
```

Comment:
This is because when we don't have an information date at the source, our information date is the processing date, and we can't run for previous information dates, hence empty Seq. But when the source has an information column, new offsets could have arrived for that specific date, so we allow it to run.

When I looked at the code, and especially when started writing unit tests for it as you suggested, I decided to refactor it a bit to make it less confising.